File: app/src/main/java/com/example/bookify/data/DatabaseHelper.java
package com.example.bookify.data;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

/**
 * Local-only data: users, bookings/tickets, and promoter applications. Events themselves live in
 * Cloud Firestore (see EventsRepository) so they sync across devices — this class no longer
 * stores or queries events directly. Bookings carry a denormalized snapshot of the event fields
 * they need (title/date/category/price/image) so ticket/check-in/revenue lookups stay local and
 * synchronous without joining across the Firestore/SQLite boundary.
 */
public class DatabaseHelper extends SQLiteOpenHelper {

    private static final String DB_NAME    = "bookify.db";
    private static final int    DB_VERSION = 11;

    // Local-only test account: bypasses Firebase Auth entirely (see LoginActivity).
    public static final String TEST_USER_EMAIL = "normal@gmail.com";
    // Sentinel firebase_uid for the local test account, so it can still own/organize events
    // (Firestore's organizerId needs some stable identifier, and this account never authenticates
    // with Firebase to get a real one). Single-device only, same as the account always was.
    public static final String TEST_USER_UID = "LOCAL_TEST_USER_UID";

    public DatabaseHelper(Context context) {
        super(context, DB_NAME, null, DB_VERSION);
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        db.execSQL("CREATE TABLE users (" +
                "id INTEGER PRIMARY KEY AUTOINCREMENT," +
                "full_name TEXT NOT NULL," +
                "email TEXT NOT NULL UNIQUE," +
                "firebase_uid TEXT UNIQUE," +
                "phone TEXT," +
                "role TEXT NOT NULL DEFAULT 'USER')");

        db.execSQL("CREATE TABLE bookings (" +
                "id INTEGER PRIMARY KEY AUTOINCREMENT," +
                "user_id INTEGER NOT NULL," +
                "event_id TEXT NOT NULL," +
                "ticket_number TEXT NOT NULL," +
                "status TEXT DEFAULT 'PENDING'," +
                "checked_in INTEGER DEFAULT 0," +
                "event_title TEXT," +
                "event_date TEXT," +
                "event_category TEXT," +
                "event_price TEXT," +
                "event_image_url TEXT," +
                "FOREIGN KEY(user_id) REFERENCES users(id))");

        db.execSQL("CREATE TABLE promoter_applications (" +
                "id INTEGER PRIMARY KEY AUTOINCREMENT," +
                "user_id INTEGER NOT NULL," +
                "hall_name TEXT NOT NULL," +
                "location TEXT NOT NULL," +
                "description TEXT," +
                "status TEXT DEFAULT 'PENDING'," +
                "latitude REAL," +
                "longitude REAL," +
                "FOREIGN KEY(user_id) REFERENCES users(id))");

        seedAdmin(db);
        seedTestUser(db);
    }

    private void seedAdmin(SQLiteDatabase db) {
        // firebase_uid is left null; linked to the Firebase Auth account on first login
        // (create admin@gmail.com / 123456 in the Firebase console under Authentication > Users)
        ContentValues cv = new ContentValues();
        cv.put("full_name", "Admin User");
        cv.put("email", "admin@gmail.com");
        cv.put("phone", "0700000000");
        cv.put("role", User.ROLE_ADMIN);
        db.insert("users", null, cv);
    }

    private void seedTestUser(SQLiteDatabase db) {
        // Local-only account, never touches Firebase — LoginActivity special-cases this
        // email and logs straight in against this row, so login always works offline.
        ContentValues cv = new ContentValues();
        cv.put("full_name", "Normal User");
        cv.put("email", TEST_USER_EMAIL);
        cv.put("firebase_uid", TEST_USER_UID);
        cv.put("phone", "0695880700");
        cv.put("role", User.ROLE_USER);
        db.insert("users", null, cv);
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        db.execSQL("DROP TABLE IF EXISTS promoter_applications");
        db.execSQL("DROP TABLE IF EXISTS bookings");
        db.execSQL("DROP TABLE IF EXISTS events");
        db.execSQL("DROP TABLE IF EXISTS users");
        onCreate(db);
    }

    // ── User methods ──────────────────────────────────────────────────────────
    // Credentials (email/password) are verified by Firebase Auth. This table only
    // holds the local profile (name, phone, role) linked by firebase_uid.

    public long registerLocalProfile(String fullName, String email, String firebaseUid, String phone) {
        ContentValues cv = new ContentValues();
        cv.put("full_name", fullName);
        cv.put("email", email);
        cv.put("firebase_uid", firebaseUid);
        cv.put("phone", phone);
        cv.put("role", User.ROLE_USER);
        return getWritableDatabase().insert("users", null, cv);
    }

    public User getUserByFirebaseUid(String uid) {
        Cursor c = getReadableDatabase().rawQuery(
                "SELECT id, firebase_uid, full_name, email, phone, role FROM users WHERE firebase_uid=?",
                new String[]{uid});
        User user = readUser(c);
        c.close();
        return user;
    }

    public User getUserByEmail(String email) {
        Cursor c = getReadableDatabase().rawQuery(
                "SELECT id, firebase_uid, full_name, email, phone, role FROM users WHERE email=?",
                new String[]{email});
        User user = readUser(c);
        c.close();
        return user;
    }

    public User getUserById(int userId) {
        Cursor c = getReadableDatabase().rawQuery(
                "SELECT id, firebase_uid, full_name, email, phone, role FROM users WHERE id=?",
                new String[]{String.valueOf(userId)});
        User user = readUser(c);
        c.close();
        return user;
    }

    public void linkFirebaseUid(int localId, String uid) {
        ContentValues cv = new ContentValues();
        cv.put("firebase_uid", uid);
        getWritableDatabase().update("users", cv, "id=?", new String[]{String.valueOf(localId)});
    }

    private User readUser(Cursor c) {
        if (c.moveToFirst()) {
            return new User(c.getInt(0), c.getString(1), c.getString(2), c.getString(3), c.getString(4), c.getString(5));
        }
        return null;
    }

    // ── Promoter application methods ────────────────────────────────────────────

    public long submitPromoterApplication(int userId, String hallName, String location, String description,
                                           Double latitude, Double longitude) {
        ContentValues cv = new ContentValues();
        cv.put("user_id", userId);
        cv.put("hall_name", hallName);
        cv.put("location", location);
        cv.put("description", description);
        cv.put("status", PromoterApplication.STATUS_PENDING);
        if (latitude != null && longitude != null) {
            cv.put("latitude", latitude);
            cv.put("longitude", longitude);
        }
        return getWritableDatabase().insert("promoter_applications", null, cv);
    }

    public PromoterApplication getLatestPromoterApplication(int userId) {
        Cursor c = getReadableDatabase().rawQuery(
                "SELECT pa.id, pa.user_id, u.full_name, u.email, pa.hall_name, pa.location, pa.description, pa.status, pa.latitude, pa.longitude " +
                "FROM promoter_applications pa JOIN users u ON pa.user_id = u.id " +
                "WHERE pa.user_id=? ORDER BY pa.id DESC LIMIT 1",
                new String[]{String.valueOf(userId)});
        PromoterApplication app = null;
        if (c.moveToFirst()) app = readPromoterApplication(c);
        c.close();
        return app;
    }

    public List<PromoterApplication> getPendingPromoterApplications() {
        List<PromoterApplication> list = new ArrayList<>();
        Cursor c = getReadableDatabase().rawQuery(
                "SELECT pa.id, pa.user_id, u.full_name, u.email, pa.hall_name, pa.location, pa.description, pa.status, pa.latitude, pa.longitude " +
                "FROM promoter_applications pa JOIN users u ON pa.user_id = u.id " +
                "WHERE pa.status=?",
                new String[]{PromoterApplication.STATUS_PENDING});
        while (c.moveToNext()) list.add(readPromoterApplication(c));
        c.close();
        return list;
    }

    private PromoterApplication readPromoterApplication(Cursor c) {
        return new PromoterApplication(c.getInt(0), c.getInt(1), c.getString(2), c.getString(3),
                c.getString(4), c.getString(5), c.getString(6), c.getString(7),
                c.isNull(8) ? null : c.getDouble(8), c.isNull(9) ? null : c.getDouble(9));
    }

    public void approvePromoterApplication(int applicationId, int userId) {
        ContentValues appCv = new ContentValues();
        appCv.put("status", PromoterApplication.STATUS_APPROVED);
        getWritableDatabase().update("promoter_applications", appCv, "id=?",
                new String[]{String.valueOf(applicationId)});

        ContentValues userCv = new ContentValues();
        userCv.put("role", User.ROLE_PROMOTER);
        getWritableDatabase().update("users", userCv, "id=?", new String[]{String.valueOf(userId)});
    }

    public void rejectPromoterApplication(int applicationId) {
        ContentValues cv = new ContentValues();
        cv.put("status", PromoterApplication.STATUS_REJECTED);
        getWritableDatabase().update("promoter_applications", cv, "id=?",
                new String[]{String.valueOf(applicationId)});
    }

    public List<PromoterProfile> getApprovedPromoters() {
        List<PromoterProfile> list = new ArrayList<>();
        Cursor c = getReadableDatabase().rawQuery(
                "SELECT u.id, u.firebase_uid, u.full_name, u.email, pa.hall_name, pa.location, pa.description, pa.latitude, pa.longitude " +
                "FROM promoter_applications pa JOIN users u ON pa.user_id = u.id " +
                "WHERE pa.status=?",
                new String[]{PromoterApplication.STATUS_APPROVED});
        while (c.moveToNext()) {
            list.add(new PromoterProfile(c.getInt(0), c.getString(1), c.getString(2), c.getString(3),
                    c.getString(4), c.getString(5), c.getString(6),
                    c.isNull(7) ? null : c.getDouble(7), c.isNull(8) ? null : c.getDouble(8)));
        }
        c.close();
        return list;
    }

    public PromoterProfile getPromoterProfile(int promoterUserId) {
        Cursor c = getReadableDatabase().rawQuery(
                "SELECT u.id, u.firebase_uid, u.full_name, u.email, pa.hall_name, pa.location, pa.description, pa.latitude, pa.longitude " +
                "FROM promoter_applications pa JOIN users u ON pa.user_id = u.id " +
                "WHERE pa.status=? AND u.id=? ORDER BY pa.id DESC LIMIT 1",
                new String[]{PromoterApplication.STATUS_APPROVED, String.valueOf(promoterUserId)});
        PromoterProfile profile = null;
        if (c.moveToFirst()) {
            profile = new PromoterProfile(c.getInt(0), c.getString(1), c.getString(2), c.getString(3),
                    c.getString(4), c.getString(5), c.getString(6),
                    c.isNull(7) ? null : c.getDouble(7), c.isNull(8) ? null : c.getDouble(8));
        }
        c.close();
        return profile;
    }

    // ── Booking methods ───────────────────────────────────────────────────────
    // eventId is now the Firestore document id (String). Event display fields are
    // denormalized onto each booking row at booking time so these stay local/synchronous.

    public int getTicketCount(String eventId, boolean completedOnly) {
        String sql = "SELECT COUNT(*) FROM bookings WHERE event_id=?";
        if (completedOnly) sql += " AND status='COMPLETED'";
        Cursor c = getReadableDatabase().rawQuery(sql, new String[]{eventId});
        int count = 0;
        if (c.moveToFirst()) count = c.getInt(0);
        c.close();
        return count;
    }

    public boolean isAlreadyBooked(int userId, String eventId) {
        Cursor c = getReadableDatabase().rawQuery(
                "SELECT id FROM bookings WHERE user_id=? AND event_id=?",
                new String[]{String.valueOf(userId), eventId});
        boolean booked = c.moveToFirst();
        c.close();
        return booked;
    }

    public long bookEvent(int userId, String eventId, String eventTitle, String eventDate,
                           String eventCategory, String eventPrice, String eventImageUrl) {
        Random rand = new Random();
        String ticket = "BKF-" + String.format("%04d", rand.nextInt(9000) + 1000)
                      + "-" + String.format("%04d", rand.nextInt(9000) + 1000);
        ContentValues cv = new ContentValues();
        cv.put("user_id", userId);
        cv.put("event_id", eventId);
        cv.put("ticket_number", ticket);
        cv.put("status", "PENDING");
        cv.put("event_title", eventTitle);
        cv.put("event_date", eventDate);
        cv.put("event_category", eventCategory);
        cv.put("event_price", eventPrice);
        cv.put("event_image_url", eventImageUrl);
        return getWritableDatabase().insert("bookings", null, cv);
    }

    public void completePayment(int userId, String eventId) {
        ContentValues cv = new ContentValues();
        cv.put("status", "COMPLETED");
        getWritableDatabase().update("bookings", cv, "user_id=? AND event_id=?",
                new String[]{String.valueOf(userId), eventId});
    }

    public String getTicketNumber(int userId, String eventId) {
        Cursor c = getReadableDatabase().rawQuery(
                "SELECT ticket_number FROM bookings WHERE user_id=? AND event_id=? AND status='COMPLETED'",
                new String[]{String.valueOf(userId), eventId});
        String ticket = null;
        if (c.moveToFirst()) ticket = c.getString(0);
        c.close();
        return ticket;
    }

    public List<Booking> getUserBookings(int userId) {
        List<Booking> list = new ArrayList<>();
        Cursor c = getReadableDatabase().rawQuery(
                "SELECT ticket_number, event_title, event_date, event_category, event_price, event_image_url, status " +
                "FROM bookings WHERE user_id=? AND status='COMPLETED'",
                new String[]{String.valueOf(userId)});
        while (c.moveToNext()) {
            list.add(new Booking(c.getString(0), c.getString(1), c.getString(2),
                    c.getString(3), c.getString(4), c.getString(5), c.getString(6)));
        }
        c.close();
        return list;
    }

    // ── Admin methods ────────────────────────────────────────────────────────

    public List<User> getAllUsers() {
        List<User> list = new ArrayList<>();
        Cursor c = getReadableDatabase().rawQuery(
                "SELECT id, firebase_uid, full_name, email, phone, role FROM users ORDER BY id DESC", null);
        while (c.moveToNext()) {
            list.add(new User(c.getInt(0), c.getString(1), c.getString(2), c.getString(3), c.getString(4), c.getString(5)));
        }
        c.close();
        return list;
    }

    public double getRevenue(String eventId) {
        Cursor c = getReadableDatabase().rawQuery(
                "SELECT event_price FROM bookings WHERE event_id=? AND status='COMPLETED'",
                new String[]{eventId});
        double total = 0;
        while (c.moveToNext()) {
            total += parsePrice(c.getString(0));
        }
        c.close();
        return total;
    }

    private double parsePrice(String price) {
        if (price == null) return 0;
        String digits = price.replaceAll("[^0-9.]", "");
        if (digits.isEmpty()) return 0;
        try {
            return Double.parseDouble(digits);
        } catch (NumberFormatException e) {
            return 0;
        }
    }

    // Admin-created promoters skip the application/approval flow, so the account
    // is inserted with the promoter role already set. The Firebase account is
    // created by the caller (AdminActivity) since FirebaseAuth calls are async.
    public long registerUserWithRole(String fullName, String email, String firebaseUid, String phone, String role) {
        ContentValues cv = new ContentValues();
        cv.put("full_name", fullName);
        cv.put("email", email);
        cv.put("firebase_uid", firebaseUid);
        cv.put("phone", phone);
        cv.put("role", role);
        return getWritableDatabase().insert("users", null, cv);
    }

    public void verifyPromoter(int userId) {
        ContentValues cv = new ContentValues();
        cv.put("role", User.ROLE_PROMOTER);
        getWritableDatabase().update("users", cv, "id=?", new String[]{String.valueOf(userId)});
    }

    /**
     * Validates a scanned ticket QR against a specific event and, on first scan,
     * marks it checked in. Reusing the same ticket_number a second time (or scanning
     * it at the wrong event) is rejected instead of silently succeeding again.
     */
    public CheckInResult checkInTicket(String ticketNumber, String eventId) {
        Cursor c = getReadableDatabase().rawQuery(
                "SELECT b.status, b.checked_in, b.event_id, u.full_name, b.event_title " +
                        "FROM bookings b " +
                        "JOIN users u ON b.user_id = u.id " +
                        "WHERE b.ticket_number=?",
                new String[]{ticketNumber});

        if (!c.moveToFirst()) {
            c.close();
            return new CheckInResult(CheckInResult.Status.INVALID_TICKET, null, null);
        }

        String status = c.getString(0);
        boolean checkedIn = c.getInt(1) != 0;
        String bookingEventId = c.getString(2);
        String attendeeName = c.getString(3);
        String eventTitle = c.getString(4);
        c.close();

        if (!bookingEventId.equals(eventId)) {
            return new CheckInResult(CheckInResult.Status.WRONG_EVENT, attendeeName, eventTitle);
        }
        if (!"COMPLETED".equals(status)) {
            return new CheckInResult(CheckInResult.Status.UNPAID, attendeeName, eventTitle);
        }
        if (checkedIn) {
            return new CheckInResult(CheckInResult.Status.ALREADY_CHECKED_IN, attendeeName, eventTitle);
        }

        ContentValues cv = new ContentValues();
        cv.put("checked_in", 1);
        getWritableDatabase().update("bookings", cv, "ticket_number=?", new String[]{ticketNumber});
        return new CheckInResult(CheckInResult.Status.CHECKED_IN, attendeeName, eventTitle);
    }

    public boolean updateUserName(int userId, String fullName) {
        ContentValues cv = new ContentValues();
        cv.put("full_name", fullName);
        int rows = getWritableDatabase().update("users", cv, "id=?", new String[]{String.valueOf(userId)});
        return rows > 0;
    }
}

File: app/src/main/java/com/example/bookify/data/DatabaseHelper.kt
package com.example.bookify.data

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper

class DatabaseHelper(context: Context) : SQLiteOpenHelper(context, "bookify.db", null, 1) {

    override fun onCreate(db: SQLiteDatabase) {
        db.execSQL("""
            CREATE TABLE users (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                fullName TEXT NOT NULL,
                email TEXT NOT NULL UNIQUE,
                password TEXT NOT NULL,
                phone TEXT NOT NULL
            )
        """)
        db.execSQL("""
            CREATE TABLE events (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                title TEXT NOT NULL,
                location TEXT NOT NULL,
                date TEXT NOT NULL,
                time TEXT NOT NULL,
                category TEXT NOT NULL,
                price TEXT NOT NULL,
                slotsLeft INTEGER NOT NULL,
                description TEXT NOT NULL,
                isPrivate INTEGER NOT NULL DEFAULT 0,
                inviteCode TEXT,
                cardColorHex TEXT NOT NULL
            )
        """)
        db.execSQL("""
            CREATE TABLE bookings (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                userId INTEGER NOT NULL,
                eventId INTEGER NOT NULL,
                ticketNo TEXT NOT NULL,
                status TEXT NOT NULL DEFAULT 'upcoming'
            )
        """)
        seedEvents(db)
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        db.execSQL("DROP TABLE IF EXISTS bookings")
        db.execSQL("DROP TABLE IF EXISTS events")
        db.execSQL("DROP TABLE IF EXISTS users")
        onCreate(db)
    }

    private fun seedEvents(db: SQLiteDatabase) {
        val events = listOf(
            ContentValues().apply {
                put("title", "Sauti Sol Live in DSM")
                put("location", "Mlimani City Ground")
                put("date", "Jun 20, 2025")
                put("time", "7:00 PM")
                put("category", "Concert")
                put("price", "Tsh 15,000")
                put("slotsLeft", 48)
                put("description", "Join us for an unforgettable night with Sauti Sol performing their greatest hits live in Dar es Salaam.")
                put("isPrivate", 0)
                put("cardColorHex", "251F5C")
            },
            ContentValues().apply {
                put("title", "DSM Tech Conference 2025")
                put("location", "Julius Nyerere CC")
                put("date", "Jun 25, 2025")
                put("time", "9:00 AM")
                put("category", "Conference")
                put("price", "Tsh 30,000")
                put("slotsLeft", 120)
                put("description", "Tanzania's premier technology conference bringing together innovators, developers, and entrepreneurs.")
                put("isPrivate", 0)
                put("cardColorHex", "1A3D2B")
            },
            ContentValues().apply {
                put("title", "DIT Alumni Gala 2025")
                put("location", "DIT Main Hall")
                put("date", "Jul 5, 2025")
                put("time", "6:00 PM")
                put("category", "Gala")
                put("price", "Tsh 20,000")
                put("slotsLeft", 60)
                put("description", "Annual DIT Alumni Gala celebrating excellence and achievements of our graduates.")
                put("isPrivate", 1)
                put("inviteCode", "DIT2025")
                put("cardColorHex", "3D1A5C")
            },
            ContentValues().apply {
                put("title", "Startup Mixer DSM")
                put("location", "Kariakoo Hub")
                put("date", "Jun 28, 2025")
                put("time", "5:00 PM")
                put("category", "Networking")
                put("price", "Tsh 10,000")
                put("slotsLeft", 30)
                put("description", "Meet the brightest startup founders and investors in Dar es Salaam at this exclusive mixer.")
                put("isPrivate", 1)
                put("inviteCode", "STARTUP28")
                put("cardColorHex", "1A3D2B")
            }
        )
        events.forEach { db.insert("events", null, it) }
    }

    fun registerUser(fullName: String, email: String, password: String, phone: String): Boolean {
        return try {
            val cv = ContentValues().apply {
                put("fullName", fullName)
                put("email", email)
                put("password", password)
                put("phone", phone)
            }
            writableDatabase.insert("users", null, cv) != -1L
        } catch (e: Exception) {
            false
        }
    }

    fun loginUser(email: String, password: String): User? {
        val cursor = readableDatabase.rawQuery(
            "SELECT * FROM users WHERE email=? AND password=?", arrayOf(email, password)
        )
        return if (cursor.moveToFirst()) {
            val user = User(
                id = cursor.getInt(cursor.getColumnIndexOrThrow("id")),
                fullName = cursor.getString(cursor.getColumnIndexOrThrow("fullName")),
                email = cursor.getString(cursor.getColumnIndexOrThrow("email")),
                phone = cursor.getString(cursor.getColumnIndexOrThrow("phone"))
            )
            cursor.close()
            user
        } else {
            cursor.close()
            null
        }
    }

    fun getAllPublicEvents(): List<Event> {
        val list = mutableListOf<Event>()
        val cursor = readableDatabase.rawQuery(
            "SELECT * FROM events WHERE isPrivate=0", null
        )
        while (cursor.moveToNext()) list.add(cursorToEvent(cursor))
        cursor.close()
        return list
    }

    fun getEventById(id: Int): Event? {
        val cursor = readableDatabase.rawQuery("SELECT * FROM events WHERE id=?", arrayOf(id.toString()))
        val event = if (cursor.moveToFirst()) cursorToEvent(cursor) else null
        cursor.close()
        return event
    }

    fun getEventByInviteCode(code: String): Event? {
        val cursor = readableDatabase.rawQuery("SELECT * FROM events WHERE inviteCode=?", arrayOf(code))
        val event = if (cursor.moveToFirst()) cursorToEvent(cursor) else null
        cursor.close()
        return event
    }

    fun bookEvent(userId: Int, eventId: Int): String? {
        val ticketNo = "BKF-${System.currentTimeMillis() % 10000000}".let {
            "BKF-2025-${(1000..9999).random()}"
        }
        val cv = ContentValues().apply {
            put("userId", userId)
            put("eventId", eventId)
            put("ticketNo", ticketNo)
            put("status", "upcoming")
        }
        val id = writableDatabase.insert("bookings", null, cv)
        if (id != -1L) {
            writableDatabase.execSQL("UPDATE events SET slotsLeft = slotsLeft - 1 WHERE id=?", arrayOf(eventId))
        }
        return if (id != -1L) ticketNo else null
    }

    fun getUserBookings(userId: Int): List<Booking> {
        val list = mutableListOf<Booking>()
        val cursor = readableDatabase.rawQuery("""
            SELECT b.id, b.ticketNo, b.status, e.title, e.date, e.time, e.category, e.cardColorHex
            FROM bookings b JOIN events e ON b.eventId = e.id
            WHERE b.userId=?
            ORDER BY b.id DESC
        """, arrayOf(userId.toString()))
        while (cursor.moveToNext()) {
            list.add(Booking(
                id = cursor.getInt(0),
                ticketNo = cursor.getString(1),
                status = cursor.getString(2),
                eventTitle = cursor.getString(3),
                eventDate = cursor.getString(4),
                eventTime = cursor.getString(5),
                eventCategory = cursor.getString(6),
                cardColorHex = cursor.getString(7)
            ))
        }
        cursor.close()
        return list
    }

    fun isAlreadyBooked(userId: Int, eventId: Int): Boolean {
        val cursor = readableDatabase.rawQuery(
            "SELECT id FROM bookings WHERE userId=? AND eventId=?",
            arrayOf(userId.toString(), eventId.toString())
        )
        val result = cursor.moveToFirst()
        cursor.close()
        return result
    }

    private fun cursorToEvent(cursor: android.database.Cursor) = Event(
        id = cursor.getInt(cursor.getColumnIndexOrThrow("id")),
        title = cursor.getString(cursor.getColumnIndexOrThrow("title")),
        location = cursor.getString(cursor.getColumnIndexOrThrow("location")),
        date = cursor.getString(cursor.getColumnIndexOrThrow("date")),
        time = cursor.getString(cursor.getColumnIndexOrThrow("time")),
        category = cursor.getString(cursor.getColumnIndexOrThrow("category")),
        price = cursor.getString(cursor.getColumnIndexOrThrow("price")),
        slotsLeft = cursor.getInt(cursor.getColumnIndexOrThrow("slotsLeft")),
        description = cursor.getString(cursor.getColumnIndexOrThrow("description")),
        isPrivate = cursor.getInt(cursor.getColumnIndexOrThrow("isPrivate")) == 1,
        inviteCode = cursor.getString(cursor.getColumnIndexOrThrow("inviteCode")),
        cardColorHex = cursor.getString(cursor.getColumnIndexOrThrow("cardColorHex"))
    )
}
